geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterElderManager.java

```diff
@@ -0,0 +1,208 @@
+
+  public ClusterElderManager(ClusterDistributionManager clusterDistributionManager) {
+    this(clusterDistributionManager, () -> new ElderState(clusterDistributionManager));
+  }
+
+  public ClusterElderManager(ClusterDistributionManager clusterDistributionManager,
+      Supplier<ElderState> elderStateSupplier) {
+    this.clusterDistributionManager = clusterDistributionManager;
+    this.elderLock = new StoppableReentrantLock(clusterDistributionManager.getCancelCriterion());
+    this.elderStateSupplier = elderStateSupplier;
+  }
+
+  /**
+   * Based on a recent JGroups view, return a member that might be the next elder.
+   *
+   * @return the elder candidate, possibly this VM.
+   */
+  InternalDistributedMember getElderCandidate() {
+    return getElderCandidates().stream().findFirst().orElse(null);
+  }
+
+  public boolean isPeerless() {
+    List<InternalDistributedMember> candidates = getElderCandidates();
+
+    return candidates.size() == 1 && clusterDistributionManager.getId()
+        .equals(candidates.get(0));
+  }
+
+  List<InternalDistributedMember> getElderCandidates() {
+    List<InternalDistributedMember> theMembers = clusterDistributionManager.getViewMembers();
+
+    return theMembers.stream()
+        .filter(member -> member.getVmKind() != ClusterDistributionManager.ADMIN_ONLY_DM_TYPE)
+        .filter(
+            member -> !clusterDistributionManager.getMembershipManager().isSurpriseMember(member))
+        .collect(Collectors.toList());
+  }
+
+  public InternalDistributedMember getElderId() throws DistributedSystemDisconnectedException {
+    if (clusterDistributionManager.isCloseInProgress()) {
+      throw new DistributedSystemDisconnectedException(
+          LocalizedStrings.DistributionManager_NO_VALID_ELDER_WHEN_SYSTEM_IS_SHUTTING_DOWN
+              .toLocalizedString(),
+          clusterDistributionManager.getRootCause());
+    }
+    clusterDistributionManager.getSystem().getCancelCriterion().checkCancelInProgress(null);
+
+    return getElderCandidate();
+  }
+
+  public boolean isElder() {
+    return clusterDistributionManager.getId().equals(getElderCandidate());
+  }
+
+  public ElderState getElderState(boolean waitToBecomeElder) {
+    if (waitToBecomeElder) {
+      // This should always return true.
+      waitForElder(clusterDistributionManager.getId());
+    }
+
+    if (clusterDistributionManager.getId().equals(getElderCandidate())) {
+      // we are the elder
+      if (this.elderStateInitialized) {
+        return this.elderState;
+      }
+      return initializeElderState();
+    } else {
+      // we are not the elder so return null
+      return null;
+    }
+  }
+
+  private ElderState initializeElderState() {
+    this.elderLock.lock();
+    try {
+      if (this.elderState == null) {
+        this.elderState = elderStateSupplier.get();
+      }
+    } finally {
+      this.elderLock.unlock();
+    }
+    this.elderStateInitialized = true;
```

Comment:
Having this line here is okay, but I think it would be even better if it was moved into the "if (this.elderState == null)" block. That way we only set it to true when we are the thread setting "elderState".